ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/editor/plugin/ExecutionEnvironmentSection.java

```diff
@@ -324,6 +350,28 @@ private void handleAdd() {
 		}
 	}
 
+	private void handleUpgrade() {
+
+		RequiredExecutionEnvironmentHeader header = getHeader();
+		IExecutionEnvironment[] executionEnvironments = JavaRuntime.getExecutionEnvironmentsManager()
+				.getExecutionEnvironments();
+		if (executionEnvironments.length > 0) {
+			ExecutionEnvironment[] addedExecutionEnvironments = header.getEnvironments();
+			// Remove existing EE ( Upgrade is allowed only when single Java EE
+			// is present in the header)
+			getHeader().removeExecutionEnvironment(addedExecutionEnvironments[0]);
+
+			SortedMap<String, IExecutionEnvironment> executionEnvironmentMap = new TreeMap<>(new VersionComparator());
+			Arrays.asList(executionEnvironments).forEach(executionEnvironment -> executionEnvironmentMap
+					.put(executionEnvironment.getId(), executionEnvironment));
+			// add highest Java EE to the header
+			header.addExecutionEnvironment(executionEnvironmentMap.get(executionEnvironmentMap.firstKey()));
```

Comment:
I don't think it isn't necessary to sort all EEs since IIRC the list of EEs `JavaRuntime.getExecutionEnvironmentsManager().getExecutionEnvironments()` is already sorted in ascending order and you can therefore just use the last element.

And in case it would be sorted, I would use
`Comparator.comparing(ee -> new org.osgi.framework.Version(getVersionString(ee)))`
to create a comparator without the need to implement a version comparison logic again.

Reply:
I won't rely on the order of some method we don't control.

Reply:
> And in case it would be sorted, I would use
> `Comparator.comparing(ee -> new org.osgi.framework.Version(getVersionString(ee)))`
> to create a comparator without the need to implement a version comparison logic again.

@HannesWell I only have option to set Comparator to `org.eclipse.ui.dialogs.FilteredList` and the sorting is done in `TwoArrayQuickSorter` .

I get the versions in ascending order using `Comparator.comparing(ee -> new org.osgi.framework.Version(getVersionString(ee)))` but the requirement is to get it in descending order .

Reply:
@HannesWell Please review the latest code